marlowe-plutus/src/Language/Marlowe/Plutus/Semantics/Types.hs

```diff
@@ -0,0 +1,523 @@
+  , ('FalseObs, 10)
+  ]
+
+-- | The (inclusive) bound on a choice number.
+data Bound = Bound Integer Integer
+  deriving stock (Haskell.Show, Generic, Haskell.Eq, Haskell.Ord)
+
+makeIsDataIndexed ''Bound [('Bound, 0)]
+
+asData
+  [d|
+    -- \| Actions happen at particular points during execution.
+    --   Three kinds of action are possible:
+    --
+    --   * A @Deposit n p v@ makes a deposit of value @v@ into account @n@ belonging to party @p@.
+    --
+    --   * A choice is made for a particular id with a list of bounds on the values that are acceptable.
+    --     For example, @[(0, 0), (3, 5]@ offers the choice of one of 0, 3, 4 and 5.
+    --
+    --   * The contract is notified that a particular observation be made.
+    --     Typically this would be done by one of the parties,
+    --     or one of their wallets acting automatically.
+    data Action
+      = Deposit AccountId Party Token (Value Observation)
+      | Choice ChoiceId [Bound]
+      | Notify Observation
+      deriving stock (Generic, Data)
+      deriving newtype (ToData, FromData, UnsafeFromData, Haskell.Eq, Haskell.Ord, Haskell.Show)
+    |]
+
+-- | A payment can be made to one of the parties to the contract,
+--   or to one of the accounts of the contract,
+--   and this is reflected in the definition.
+data Payee
+  = Account AccountId
+  | Party Party
+  deriving stock (Haskell.Show, Generic, Haskell.Eq, Haskell.Ord)
+
+makeIsDataIndexed ''Payee [('Account, 0), ('Party, 1)]
+
+-- | A case is a branch of a when clause, guarded by an action.
+--   The continuation of the contract may be merkleized or not.
+--
+--   Plutus doesn't support mutually recursive data types yet.
+--   datatype Case is mutually recursive with @Contract@
+data Case a
+  = Case Action a
+  | MerkleizedCase Action BuiltinByteString
+  deriving stock (Haskell.Show, Generic, Haskell.Eq, Haskell.Ord)
+
+makeIsDataIndexed ''Case [('Case, 0), ('MerkleizedCase, 1)]
+
+-- | Extract the @Action@ from a @Case@.
+getAction :: Case a -> Action
+getAction (Case action _) = action
+getAction (MerkleizedCase action _) = action
+{-# INLINEABLE getAction #-}
+
+asData
+  [d|
+    -- \| Marlowe has six ways of building contracts.
+    --   Five of these – 'Pay', 'Let', 'If', 'When' and 'Assert' –
+    --   build a complex contract from simpler contracts, and the sixth, 'Close',
+    --   is a simple contract.
+    --   At each step of execution, as well as returning a new state and continuation contract,
+    --   it is possible that effects – payments – and warnings can be generated too.
+    data Contract
+      = Close
+      | Pay AccountId Payee Token (Value Observation) Contract
+      | If Observation Contract Contract
+      | When [Case Contract] Timeout Contract
+      | Let ValueId (Value Observation) Contract
+      | Assert Observation Contract
+      deriving stock (Generic, Data)
+      deriving newtype (ToData, FromData, UnsafeFromData, Haskell.Eq, Haskell.Ord, Haskell.Show)
+    |]
```

Comment:
Also for `Action`

Reply:
Using `asData` makes `makeIsDataIndexed` redundant because the patterns include that information. I'll double-check the TH source code for `makeIsDataIndexed`, just to be 100% sure.

Reply:
> In the same line, we could order things in marlowe-cardano in the same way than in marlowe-plutus.

If you're referring to the ordering in the semantics files, we'd have to change those in `marlowe-cardano` because the order in `marlowe-plutus` is (unfortunately) constrained by the way GHC handles TH: i.e., identifiers don't resolve during compilation of the module unless the ordering declares them in a particular sequence.

I think we should reorder and clean up the `marlowe` module in `marlowe-cardano`, maybe when we upgrade to Conway near the start of PI5. Typeclass instances etc. are scattered about the file in a not very logical order. @jhbertra has some ideas for improvements.

Reply:
Ah, if it is not possible to be explicit in `marlowe-plutus` then it makes sense

Reply:
@bwbush isn't `makeIsDataIndexed` replaced in the TS block by `deriving stock (Data,...)` which is "constructor order dependent" derivation of the encoding instance:
* `Action`: https://github.com/input-output-hk/marlowe-plutus/pull/12/files#diff-fcfa0434b7f0a95ec914bd85cc1b722b562d81c2d676c51768a9de4b9e1c0b77R246
* `Contract`: https://github.com/input-output-hk/marlowe-plutus/pull/12/files#diff-fcfa0434b7f0a95ec914bd85cc1b722b562d81c2d676c51768a9de4b9e1c0b77R293

Reply:
Actually, `makeIsDataIndexed` is replaced by `asData`, whose splice expands to explicitly numbering the constructors in their order of appearance. The `deriving stock Data` refers to the type `Data.Data (Data)` in the base library and doesn't particularly relate to Plutus.

Reply:
> Ah, if it is not possible to be explicit in marlowe-plutus then it makes sense

I've added a comment documenting this in 392246d0537e80c9c552956282d1569401f351c0.